src/core/exec-invoke.c

```diff
@@ -2548,6 +2557,169 @@ static int create_many_symlinks(const char *root, const char *source, char **sym
+                if (r < 0)
+                        log_warning_errno(r, "Exec quotas: Failed to release project ID %" PRIu32 ", ignoring: %m", proj_id);
+        }
+
+        return 0;
+}
+
+static int apply_exec_quotas(const char *target_dir, const char *cgroup_path, ExecDirectoryType type, const QuotaLimit *ql) {
+        assert(target_dir);
+        assert(ql);
+
+        _cleanup_close_ int fd = -EBADF;
+        int quota_supported = 0;
+        int r;
+
+        if (!IN_SET(type, EXEC_DIRECTORY_STATE, EXEC_DIRECTORY_CACHE, EXEC_DIRECTORY_LOGS))
+                return 0;
+
+        fd = xopenat(AT_FDCWD, target_dir, O_RDONLY|O_CLOEXEC|O_NOCTTY|O_NOFOLLOW);
+        if (fd < 0)
+                return log_debug_errno(fd, "Exec quotas: Failed to open %s: %m", target_dir);
+
+        uint32_t proj_id;
+        r = get_proj_id(fd, &proj_id);
+        if (ERRNO_IS_NEG_IOCTL_NOT_SUPPORTED(r)) {
+                log_warning("Not applying storage quotas. Ioctl not supported for %s", target_dir);
+                return 0;
+        }
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to retrieve project ID for %s: %m", target_dir);
+
+        bool proj_id_exists = PROJ_ID_MIN <= proj_id && proj_id <= PROJ_ID_MAX;
+
+        if (!ql->quota_enforce) {
+                if (proj_id_exists) {
+                        r = unset_exec_storage_quota(fd, proj_id, ql->quota_accounting);
+                        if (r < 0)
+                                return log_debug_errno(r, "Exec quotas: Failed to unset project quotas for %s: %m", target_dir);
+                }
+
+                if (!ql->quota_accounting)
+                        return r;
+        }
+
+        if (!proj_id_exists) {
+                static const sd_id128_t k = SD_ID128_ARRAY(e1,4a,79,9b,64,40,41,4a,a8,46,c2,f3,f9,19,4f,01);
+                _cleanup_free_ char *proj_id_plain = NULL;
+
+                /* Generate candidate project id */
+                proj_id_plain = strjoin(cgroup_path, ".", exec_directory_type_to_string(type));
+                if (!proj_id_plain)
+                        return log_oom_debug();
+
+                proj_id = PROJ_ID_CLAMP_INTO_QUOTA_RANGE(siphash24(proj_id_plain, strlen(proj_id_plain), k.bytes));
+
+#define MAX_PROJ_ID_RETRIES 10
+                for (unsigned attempt = 0;; attempt++) {
+                        if (attempt >= MAX_PROJ_ID_RETRIES)
+                                return log_debug_errno(SYNTHETIC_ERRNO(EBUSY), "Exec quotas: Failed to generate unique project ID for %s: %m", target_dir);
+
+                        /* Check if project quotas are supported */
+                        struct dqblk req;
+                        quota_supported = is_proj_id_quota_supported(fd, proj_id, &req);
+                        if (quota_supported < 0)
+                                return log_debug_errno(quota_supported, "Exec quotas: Failed to query disk quota for project ID %" PRIu32 ": %m", proj_id);
+
```

Comment:
please remove empty line. we generally try to keep function calls and their immediate return/error value handling together. i.e. place empty lines before func calls, and after handling all ret vals/errors, but not in the middle of retval/error handling